fuzz/fuzz_format_parsers.c

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright (C) 2020 Yubico AB - See COPYING
+ */
+#include <string.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+#include "util.c"
+#include "b64.c"
+
+/* wrap some functions so we can let them fail some times to trigger error
+ * paths. compile with -Wl,--wrap=strdup -Wl,--wrap=calloc to activate.
+ * Note: idea taken from libfido2/fuzz/wrap.c
+ */
+extern char *__wrap_strdup(const char *);
+extern char *__real_strdup(const char *);
+char *__wrap_strdup(const char *s) {
+  if (random() < (RAND_MAX / 100) / 4) { // fail 0.25% of the time
+    errno = ENOMEM;
+    return NULL;
+  }
+
+  return __real_strdup(s);
+}
+
+extern void *__wrap_calloc(size_t, size_t);
+extern void *__real_calloc(size_t, size_t);
+void *__wrap_calloc(size_t nmemb, size_t size) {
+  if (random() < (RAND_MAX / 100) / 4) { // fail 0.25% of the time
+    errno = ENOMEM;
+    return NULL;
+  }
+
+  return __real_calloc(nmemb, size);
+}
+
+static void cleanup(device_t *devs, unsigned int n_devs) {
+  for (int i = 0; i < n_devs; i++) {
+    free(devs[i].keyHandle);
+    free(devs[i].publicKey);
+    free(devs[i].coseType);
+    free(devs[i].attributes);
+    devs[i].keyHandle = NULL;
+    devs[i].publicKey = NULL;
+    devs[i].coseType = NULL;
+    devs[i].attributes = NULL;
+  }
+}
+
+#define DEV_MAX_SIZE 10
+
+int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size) {
+  char buf[DEVSIZE * DEV_MAX_SIZE]; /* DEVSIZE * cfg.max_size */
+  device_t devs[12] = {0};
+  unsigned int n_devs = 12;
+  FILE *fp = NULL;
+  size_t fp_len = 0;
+  size_t offset = 0;
+  char username[256] = {0};
+  size_t username_len = 0;
+  uint8_t ssh_format = 1;
+  cfg_t cfg = {0};
+  cfg.max_devs = DEV_MAX_SIZE;
+
+  /* first byte decides which parser we should call so we need
+   * atleast one byte
+   */
+  if (size < 1) {
+    return -1;
+  }
+  /* do not always run with debug, only 8/255 times */
+  if (data[offset] < 9) {
+    cfg.debug = 1;
+  }
+
+  /* predictable random for this seed */
+  srandom((unsigned int) data[offset]);
+
+  /* choose which format parser to run, even == native, odd == ssh */
+  if (data[offset++] % 2) {
```

Comment:
i strongly recommend making the seed `sizeof(unsigned int)` bytes long (2^8 vs. 2^32 distinct sequences of strdup/calloc events) and separate from the rest of the corpus (so it can be independently rotated for a given code path).

Reply:
You are right, I bumped the seed to 2^32 and got better coverage. Thanks!

Not sure it is worth separating out the seed from the rest of the corpus, in this PR at least. For now I just rewrote the existing corpora files with the additional seed size change.